File: ios/Classes/CacheManager.swift
import Foundation
import CommonCrypto

// =============================================================================
// CacheManager — atomic cache directory layout
// =============================================================================
//
// Each cache entry lives in its own directory:
//
//   <root>/<sha16>/
//     ├── chunk_0.jpg
//     ├── chunk_1.jpg
//     └── meta.json
//
// Writes happen in a sibling temp directory and are renamed into place once
// the entire chunk set is committed. Readers always see complete sets — no
// .complete sentinel needed because the directory itself is the commit unit.
//
// Cache-key collision note: SHA-256(URL) truncated to 16 hex = 64 bits.
// Birthday collision is ~2^32, fine for a per-app cache. Cache invalidation
// is handled via ETag/Last-Modified, not the key itself.
// =============================================================================

final class CacheManager {

    let rootDirectory: String

    init(rootDirectory: String) {
        self.rootDirectory = rootDirectory
        try? FileManager.default.createDirectory(
            atPath: rootDirectory,
            withIntermediateDirectories: true
        )
    }

    func key(for source: String) -> String {
        return String(sha256(source).prefix(16))
    }

    func directory(for key: String) -> String {
        return (rootDirectory as NSString).appendingPathComponent(key)
    }

    /// Allocates a fresh temp directory for in-progress writes. The caller
    /// MUST commit it via `commit(_:key:)` or discard it via `discard(_:)`.
    func newTempDirectory(for key: String) throws -> String {
        let temp = (rootDirectory as NSString)
            .appendingPathComponent(".tmp_\(key)_\(Int(Date().timeIntervalSince1970 * 1000))")
        let fm = FileManager.default
        if fm.fileExists(atPath: temp) {
            try? fm.removeItem(atPath: temp)
        }
        try fm.createDirectory(atPath: temp, withIntermediateDirectories: true)
        return temp
    }

    /// Atomically moves [tempDir] into the canonical location for [key].
    /// On the same volume this is a single rename(2) syscall.
    func commit(tempDir: String, key: String) throws -> String {
        let finalDir = directory(for: key)
        let fm = FileManager.default
        if fm.fileExists(atPath: finalDir) {
            try fm.removeItem(atPath: finalDir)
        }
        try fm.moveItem(atPath: tempDir, toPath: finalDir)
        return finalDir
    }

    func discard(tempDir: String) {
        try? FileManager.default.removeItem(atPath: tempDir)
    }

    func read(key: String) -> CachedEntry? {
        let dir = directory(for: key)
        let fm = FileManager.default
        var isDir: ObjCBool = false
        guard fm.fileExists(atPath: dir, isDirectory: &isDir), isDir.boolValue else {
            return nil
        }
        let metaPath = (dir as NSString).appendingPathComponent("meta.json")
        guard let meta = MetaFile.read(from: metaPath) else { return nil }

        let chunkPaths: [String] = (0..<meta.chunkHeights.count).map {
            (dir as NSString).appendingPathComponent("chunk_\($0).jpg")
        }
        for path in chunkPaths {
            if !fm.fileExists(atPath: path) { return nil }
        }
        return CachedEntry(paths: chunkPaths, meta: meta)
    }

    @discardableResult
    func clearAll() throws -> Int64 {
        let fm = FileManager.default
        guard fm.fileExists(atPath: rootDirectory) else { return 0 }
        var count: Int64 = 0
        let enumerator = fm.enumerator(atPath: rootDirectory)
        while let entry = enumerator?.nextObject() as? String {
            let full = (rootDirectory as NSString).appendingPathComponent(entry)
            var isDir: ObjCBool = false
            if fm.fileExists(atPath: full, isDirectory: &isDir), !isDir.boolValue {
                count += 1
            }
        }
        try fm.removeItem(atPath: rootDirectory)
        try fm.createDirectory(atPath: rootDirectory, withIntermediateDirectories: true)
        return count
    }

    private func sha256(_ string: String) -> String {
        let data = Data(string.utf8)
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

struct CachedEntry {
    let paths: [String]
    let meta: CacheMeta
}

File: android/src/main/kotlin/io/github/hrjy6278/image_splitter/CacheManager.kt
package io.github.hrjy6278.image_splitter

import java.io.File
import java.security.MessageDigest

// =============================================================================
// CacheManager
// =============================================================================
//
// Atomic cache directory layout:
//
//   <cacheRoot>/image_splits/<sha16>/
//     ├── chunk_0.jpg
//     ├── chunk_1.jpg
//     ├── ...
//     └── meta.json    (etag, last-modified, chunk heights, image width)
//
// Atomicity: writes happen in a sibling temp directory and are renamed into
// place once the entire chunk set is committed. A reader that sees the
// final directory is guaranteed to see a complete set — no .complete sentinel
// needed because the directory itself is the commit unit.
//
// Cache-key collision note: SHA-256(URL) truncated to 16 hex chars = 64 bits.
// Birthday-collision probability is ~2^32, which is fine for a per-app cache
// but not for adversarial input. Cache invalidation is handled by ETag /
// Last-Modified, not the key itself.
// =============================================================================

internal class CacheManager(private val rootDir: File) {

    init {
        rootDir.mkdirs()
    }

    fun keyFor(source: String): String = sha256(source).take(16)

    fun directoryFor(key: String): File = File(rootDir, key)

    /**
     * Allocates a fresh temp directory for in-progress writes. The caller
     * MUST commit it via [commit] (renames into place) or discard it via
     * [discard] (removes it).
     */
    fun newTempDirectory(key: String): File {
        val temp = File(rootDir, ".tmp_${key}_${System.nanoTime()}")
        if (temp.exists()) temp.deleteRecursively()
        temp.mkdirs()
        return temp
    }

    /**
     * Atomically promotes [tempDir] to the canonical location for [key].
     * If a previous cache exists, it is replaced. Returns the final directory.
     */
    fun commit(tempDir: File, key: String): File {
        val finalDir = directoryFor(key)
        if (finalDir.exists()) finalDir.deleteRecursively()
        // File.renameTo on the same filesystem is atomic on POSIX.
        if (!tempDir.renameTo(finalDir)) {
            // Fallback: copy + delete (rare; happens across filesystems).
            tempDir.copyRecursively(finalDir, overwrite = true)
            tempDir.deleteRecursively()
        }
        return finalDir
    }

    fun discard(tempDir: File) {
        if (tempDir.exists()) tempDir.deleteRecursively()
    }

    fun read(key: String): CachedEntry? {
        val dir = directoryFor(key)
        if (!dir.isDirectory) return null
        val meta = MetaFile.read(File(dir, "meta.json")) ?: return null
        val chunkPaths = meta.chunkHeights.indices
            .map { File(dir, "chunk_$it.jpg") }
        if (chunkPaths.any { !it.isFile }) return null
        return CachedEntry(
            paths = chunkPaths.map { it.absolutePath },
            meta = meta
        )
    }

    fun clearAll(): Long {
        if (!rootDir.exists()) return 0
        var count = 0L
        rootDir.walkBottomUp().forEach { f ->
            if (f.isFile) count++
            f.delete()
        }
        rootDir.mkdirs()
        return count
    }

    private fun sha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(input.toByteArray())
            .joinToString("") { "%02x".format(it) }
    }
}

internal data class CachedEntry(
    val paths: List<String>,
    val meta: CacheMeta,
)
